architxt/database/__init__.py

```diff
@@ -0,0 +1,240 @@
+from sqlalchemy import create_engine, inspect, text
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.orm import sessionmaker
+
+from architxt.tree import NodeLabel, NodeType, Tree
+
+
+def read_database(db_path: str, db_type: str) -> None:
+    """
+    Read the database, retrieve table information, process table relations, and print the results in a tree format.
+
+    :param db_path: Path to the database.
+    :param db_type: Type of the database.
+    """
+    session = None
+    try:
+        # Create a connection to the SQLite database
+        engine = create_engine(f'{db_type}:///{db_path}')
+        inspector = inspect(engine)
+        session = sessionmaker(bind=engine)()
+
+        # Retrieve the list of tables in the database
+        tables = inspector.get_table_names() or []
+
+        # Get information about the tables (primary keys, foreign keys)
+        table_info = read_info_database(tables, inspector)
+
+        # Retrieve primary keys and foreign key relations
+        primary_keys = table_info.get("Primary keys", {})
+        table_relations = table_info.get("Foreign keys", {})
+        parent_tables = get_parent_tables(table_relations, tables)
+
+        final_forest = []
+        for table in parent_tables:
+            final_forest.extend(process_table_relations(table, table_relations, session, inspector, primary_keys))
+
+        for tree in final_forest:
+            print(tree.pformat(margin=255), end="\n\n\n")
+            print(tree)
+
+    except SQLAlchemyError as e:
+        print(f"SQL Error: {e}")
+    finally:
+        if session:
+            session.close()
+
+
+def read_info_database(
+    tables: list[str],
+    inspector: inspect,
+) -> dict[str, dict[str, str]]:
+    """
+    Retrieve information about the database tables, including primary and foreign keys.
+
+    :param tables: List of table names to inspect.
+    :param inspector: Database inspection object.
+    :return: A dictionary containing primary keys and foreign keys for each table.
+    """
+    table_info = {"Primary keys": {}, "Foreign keys": {}}
+
+    for table in tables:
+        primary_keys = inspector.get_pk_constraint(table)
+        table_info["Primary keys"][table] = primary_keys['constrained_columns']
+
+        foreign_keys = inspector.get_foreign_keys(table)
+        if foreign_keys:
+            table_info["Foreign keys"][table] = [
+                {'column': column, 'referred_table': fk['referred_table']}
+                for fk in foreign_keys
+                for column in fk['constrained_columns']
+            ]
+    return table_info
```

Comment:
Instead of directly manipulating dictionaries, consider refactoring this code to use SQLAlchemy reflection via the MetaData object. See: https://docs.sqlalchemy.org/en/20/core/reflection.html#reflecting-all-tables-at-once.